### PR TITLE
fix: implement constant evaluation for all expressions

### PR DESCRIPTION
**Pull request by niklasdewally**
Fixes: #481

### CODE REVIEW

File: crates/conjure_core/src/rules/constant.rs [hunk 1]
```diff
@@ -81,15 +81,50 @@ pub fn eval_constant(expr: &Expr) -> Option<Lit> {
...
+            // modulo always rounds down
+            let a = unwrap_atom::<i32>(a)?;
+            let b = unwrap_atom::<i32>(b)?;
+            let c = unwrap_atom::<i32>(c)?;
+
+            if b == 0 {
+                return None;
+            }
+
+            let modulo = (a as f32 % b as f32).floor() as i32;
```
Comment (ozgurakgun) [inline crates/conjure_core/src/rules/constant.rs:105]: this is unlikely to be correct, say for negative numbers? SR docs suggest `a-b*floor(a/b)`

Comment (niklasdewally) [reply crates/conjure_core/src/rules/constant.rs]: I added this, but it breaks the mod_02 test:

```
find a : int(0..3)
such that a >= 4 % 2 $ should be simplified to constant
```

```
Diff < left / right > :
<Array []
>Array [
>    Object {
>        "a": Number(0),
>    },
>    Object {
>        "a": Number(1),
>    },
>    Object {
>        "a": Number(2),
>    },
>    Object {
>        "a": Number(3),
>    },
>
```

Comment (niklasdewally) [reply crates/conjure_core/src/rules/constant.rs]: I see my mistake !

Comment (ozgurakgun) [reply crates/conjure_core/src/rules/constant.rs]: hmmmm, I'll look into it.

Comment (niklasdewally) [reply crates/conjure_core/src/rules/constant.rs]: I fixed it: I did (a-b) * floor(a/b) but it should be a - b * floor(a/b)

Comment (ozgurakgun) [reply crates/conjure_core/src/rules/constant.rs]: I was getting worried :) Good to see the testing infrastructure is doing its job though!

**Comment by niklasdewally** [comment]
Savile Row has some good tests for `/` and `%` constant evaluation; I am working towards adding these, but they require negation first: #490  #489

**Comment by niklasdewally** [comment]
Seems like the `fixup!` CI is confused by my merge commits.
This is ready to review, but don't merge this until I've done some commit cleanup.

File: crates/conjure_core/src/rules/constant.rs [hunk 2]
```diff
@@ -81,15 +82,50 @@ pub fn eval_constant(expr: &Expr) -> Option<Lit> {
...
 
         Expr::Reify(_, a, b) => bin_op::<bool, bool>(|a, b| a == b, a, b).map(Lit::Bool),
-        _ => {
-            warn!(%expr,"Unimplemented constant eval");
-            None
+        Expr::SumEq(_, exprs, a) => {
+            flat_op::<i32, bool>(|e, a| e.iter().sum::<i32>() == a, exprs, a).map(Lit::Bool)
+        }
+        Expr::ModuloEqUndefZero(_, a, b, c) => {
+            // modulo always rounds down
```
Comment (ozgurakgun) [inline crates/conjure_core/src/rules/constant.rs:97]: I don't think this comment is correct. Modulo works in a way that is compatible with an always-rounding-down div.

File: crates/conjure_core/src/rules/constant.rs [hunk 3]
```diff
@@ -81,15 +82,50 @@ pub fn eval_constant(expr: &Expr) -> Option<Lit> {
...
+            let a = unwrap_atom::<i32>(a)?;
+            let b = unwrap_atom::<i32>(b)?;
+            let c = unwrap_atom::<i32>(c)?;
+
+            if b == 0 {
+                return None;
+            }
+
+            let modulo = a - b * (a as f32 / b as f32).floor() as i32;
```
Comment (ozgurakgun) [inline crates/conjure_core/src/rules/constant.rs:106]: Maybe a comment before this line to say why it isn't rust's %

**Review by ozgurakgun — APPROVED** [review]

**Pull request merged**